lib/validation.js

```diff
@@ -108,9 +140,39 @@ function isValidLogger (logger) {
   return result
 }
 
-function schemaCompiler (schema) {
-  return this.ajv.compile(schema)
+function schemaCompiler (schema, keyRef) {
+  const schemaId = keyRef || id(schema)
+  return getCachedSchema(schemaId, schema, this.ajv)
+}
+
+// skips the first argument of the id since it is not needed
+function createStringifier (_, schema, root, allSchemas) {
```

Comment:
what is `root`?

Reply:
I've added comments for what each of the arguments are, including root.

If the schema referenced is inside another document (referencing 'defs.json#/definitions/test') this is that root document, which might contain other schemas that it also references. This would only come up if you added a schema through options that had definitions in it - I can add a test to cover that scenario if you'd like.